javascript/76-Minimum-Window-Substring.js

```diff
@@ -1,35 +1,78 @@
-var minWindow = function (str, target) {
-  const hash = target.split('').reduce((acc, val) => {
-    if (!acc[val]) acc[val] = 0;
-    acc[val] += 1;
-    return acc;
-  }, {})
-
-  let start = 0;
-  let min = Infinity;
-  let matched = 0;
-  let subStringStart = null;
-
-  for (let i = 0; i < str.length; i++) {
-    let rightChar = str[i];
-
-    if (rightChar in hash) hash[rightChar] -= 1;
-    if (hash[rightChar] >= 0) matched += 1;
-
-    while (matched === target.length) {
-      if (i - start + 1 < min) {
-        subStringStart = start;
-        min = i - start + 1;
-      }
-
-      let leftChar = str[start];
-      start += 1;
-
-      if (leftChar in hash) {
-        if (hash[leftChar] === 0) matched -= 1;
-        hash[leftChar] += 1;
-      }
+/**
+ * @param {string} s
+ * @param {string} t
+ * Time O(N + M) | SpaceO(N + M)
+ * @return {string}
+ */
+var minWindow = function (s, t) {
+    const isMissingArgs = !s.length || !t.length
+    if (isMissingArgs) return '';
+
+    const frequencyMap = getFrequencyMap(t);
+    const { start, end } = getWindowPointers(s, t, frequencyMap);
+
+    return getSubString(s, start, end);
+}
+
+const getFrequencyMap = (str, frequencyMap = new Map()) => {
+    for (const char of str) {
+        frequencyMap.set(char, (frequencyMap.get(char) || 0) + 1);
+    }
+
+    return frequencyMap;
+};
+
+const getWindowPointers = (s, t, frequencyMap) => {
+    let [ left, right, matched, start, end ] = [ 0, 0, 0, 0, (s.length + 1) ];
+    
+    while (right < s.length) {
+        matched = addRightFrequency(s, right, frequencyMap, matched);
+
+        const canSlide = () => matched === t.length;
+        while (canSlide()) {
```

Comment:
nit: prefer boolean to lambda here

Reply:
@mitchellirvin I mentioned the same issue in my other PR. Won't be able to a const in a while loop.